File: matlab/fastsum/fastsum.m

% Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
%
% This program is free software; you can redistribute it and/or modify it under
% the terms of the GNU General Public License as published by the Free Software
% Foundation; either version 2 of the License, or (at your option) any later
% version.
%
% This program is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
% FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
% details.
%
% You should have received a copy of the GNU General Public License along with
% this program; if not, write to the Free Software Foundation, Inc., 51
% Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

% This class provides a Matlab interface to the FASTSUM application.
classdef fastsum < handle

properties(Dependent=true)
	x           % source nodes in d-ball with radius 1/4-eps_b/2 (real Nxd matrix)
	alpha       % fourier coefficients (column vector of length N)
	y           % target nodes in d-ball with radius 1/4-eps_b/2 (real Mxd matrix)
end %properties

properties(Dependent=true,SetAccess='private')
    f;          % function evaluations in y
    b;          % expansion coefficients
    N;          % number of source nodes
end %properties

properties(SetAccess='private')
	d               % spatial dimension 
    kernel          % kernel
    c               % kernel parameter
    flags  = 0      % flags
    n               % expansion degree (of NFFT)
    p               % degree of smoothness of regularization
    eps_I           % inner boundary
    eps_B           % outer boundary
    nn_x            % oversampled nn in x
    nn_y            % oversampled nn in y
    m_x             % NFFT-cutoff in x
    m_y             % NFFT-cutoff in y
end %properties

properties(Hidden=true,SetAccess='private',GetAccess='private')
	plan
	x_is_set=false      % flag if x is set
	y_is_set=false      % flag if y is set
	alpha_is_set=false  % flag if alpha is set
	f_is_set=false      % flag if f is set
	plan_is_set=false   % flag if plan was created
end %properties

methods

% Constructer and destructor %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function h=fastsum(d,kernel,c,flags,n,p,eps_I,eps_B,nn_x,m_x,nn_y,m_y)
% Constructor
%
% h=fastsum(d,kernel,c,flags,n,p,eps_I,eps_B)
% h=fastsum(d,kernel,c,flags,n,p,eps_I,eps_B,nn,m)
% h=fastsum(d,kernel,c,flags,n,p,eps_I,eps_B,nn_x,m_x,nn_y,m_y)
%
% INPUT
% kernel = 'multiquadric', etc. (see options below)
% c       kernel parameter
% n       expansion degree
% m       cut-off parameter for NFFT
% p       degree of smoothness of regularization
% eps_I   inner boundary
% eps_B   outer boundary
% nn      oversampled nn in NFFT (in x and y)
% m       cut-off parameter for NFFT (in x and y)
% nn_x    oversampled nn in NFFT (in x)
% m_x     cut-off parameter for NFFT (in x)
% nn_y    oversampled nn in NFFT (in y)
% m_y     cut-off parameter for NFFT (in y)
%
% OUTPUT
%   h   object of class type fastsum
    narginchk(8,12);

	h.d = d;
    h.kernel = kernel;
    h.c = c;
    h.flags = flags;
    h.n = n;
    h.p = p;
    h.eps_I = eps_I;
    h.eps_B = eps_B;
    
    switch nargin
    case 8
        h.nn_x = 2*n;   % default oversampling factor 2
        h.nn_y = 2*n;
        h.m_x = p;      % default NFFT-cutoff
        h.m_y = p;
    case 10
        h.nn_x=nn_x;
        h.nn_y=nn_x;
        h.m_x=m_x;
        h.m_y=m_x;
    case 12
        h.nn_x=nn_x;
        h.nn_y=nn_y;
        h.m_x=m_x;
        h.m_y=m_y;
    otherwise
        error('Wrong number of arguments')
    end
    
    if nn_x<=h.n || nn_x<=h.n
        error('Oversampled nn must be larger than n')
    end
    
    h.plan=fastsummex('init',h.d,h.kernel,h.c,h.flags,h.n,h.p,h.eps_I,h.eps_B);
    h.plan_is_set=true;
end %function

function delete(h)
% Destructor
	if(h.plan_is_set)
		fastsummex('finalize',h.plan);
		h.plan_is_set=false;
	end %if
end %function

% Set functions %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function set.x(h,x)
    if (size(x,2)~=h.d || ~isreal(x))
        error('x must be a real matrix with d columns');
    end
    if h.x_is_set && h.alpha_is_set && size(x,1)~=h.N
        warning('fastsum:alphaDeleted',...
            'Size of x has changed, so alpha is deleted');
        h.alpha_is_set=false;
    end
    fastsummex('set_x',h.plan,x,h.nn_x,h.m_x)
    h.x_is_set = true;
end %function

function set.alpha(h,alpha)
    if ~h.x_is_set
        error('You have to set x before you set alpha')
    end
    if (~iscolumn(alpha) || ~isnumeric(alpha) || length(alpha)~=h.N)
        error('alpha must be a column vector of length N');
    end
    fastsummex('set_alpha',h.plan,alpha)
    h.alpha_is_set=true;
end %function

function set.y(h,y)
    if (size(y,2)~=h.d || ~isreal(y))
        error('y must be a real matrix with d columns');
    end
    fastsummex('set_y',h.plan,y,h.nn_y,h.m_y)
    h.y_is_set = true;
end %function

function set.d(h,d)
    if (~isscalar(d) || ~isreal(d) || d<=0 || mod(d,1)~=0)
        error('d must be a positive integer')
    end
    h.d = d;
end %function

function set.flags(h,n)
    if (~isscalar(n) || ~isreal(n) || n<0 || mod(n,1)~=0)
        error('flags must be a non-negative integer')
    end
    h.flags = n;
end %function

function set.n(h,n)
    if (~isscalar(n) || ~isreal(n) || n<=0 || mod(n,2)~=0)
        error('n must be a positive, even integer')
    end
    h.n = n;
end %function

function set.p(h,p)
    if (~isscalar(p) || ~isreal(p) || p<=0 || mod(p,1)~=0)
        error('p must be a positive integer')
    end
    h.p = p;
end %function

function set.eps_I(h,arg)
    if (~isscalar(arg) || arg<0 || arg>=pi/4)
        error('eps_I must be in the interval [0,pi/4)')
    end
    h.eps_I = arg;
end %function

function set.eps_B(h,arg)
    if (~isscalar(arg) || arg<0 || arg>=pi/4)
        error('eps_B must be in the interval [0,pi/4)')
    end
    h.eps_B = arg;
end %function

function set.nn_x(h,arg)
    if (~isscalar(arg) || ~isreal(arg) || arg<=0 || mod(arg,1)~=0)
        error('nn_x must be a positive integer')
    end
    h.nn_x = arg;
end %function

function set.m_x(h,arg)
    if (~isscalar(arg) || ~isreal(arg) || arg<=0 || mod(arg,1)~=0)
        error('m_x must be a positive integer')
    end
    h.m_x = arg;
end %function

function set.nn_y(h,arg)
    if (~isscalar(arg) || ~isreal(arg) || arg<=0 || mod(arg,1)~=0)
        error('nn_y must be a positive integer')
    end
    h.nn_y = arg;
end %function

function set.m_y(h,arg)
    if (~isscalar(arg) || ~isreal(arg) || arg<=0 || mod(arg,1)~=0)
        error('m_y must be a positive integer')
    end
    h.m_y = arg;
end %function


% Get functions %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function x=get.x(h)
    if h.x_is_set
        x=fastsummex('get_x',h.plan);
    else
        x=[];
    end
end %function

function alpha=get.alpha(h)
    if h.alpha_is_set
        alpha=fastsummex('get_alpha',h.plan);
    else
        alpha=[];
    end
end %function

function y=get.y(h)
    if h.y_is_set
        y=fastsummex('get_y',h.plan);
    else
        y=[];
    end
end %function

function b=get.b(h)
	if(h.plan_is_set)
		b=fastsummex('get_b',h.plan);
	else
		b=[];
	end %if
end %function

function f=get.f(h)
	if(h.f_is_set)
		f=fastsummex('get_f',h.plan);
	else
		f=[];
	end %if
end %function

function N=get.N(h)
	if(h.x_is_set)
		N=fastsummex('get_N_total',h.plan);
	else
		N=[];
	end %if
end %function

% User methods %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function fastsum_trafo(h)
% NFFT.fastsum
%
% fastsum_trafo(h)
%
% INPUT
%   h  object of class type nfft
    if(~h.x_is_set)
		error('Before doing a fastsum transform you have to set nodes x.');
    end
    if(~h.alpha_is_set)
		error('Before doing a fastsum transform you have to set coefficients alpha.');
    end
    if(~h.y_is_set)
		error('Before doing a fastsum transform you have to set nodes y.');
    end
        
	fastsummex('trafo',h.plan);
	h.f_is_set=true;
end %function

function fastsum_trafo_direct(h)
% NFFT.fastsum
%
% fastsum_trafo_direct(h)
%
% INPUT
%   h  object of class type nfft
    if(~h.x_is_set)
		error('Before doing a fastsum transform you have to set nodes x.');
    end
    if(~h.alpha_is_set)
		error('Before doing a fastsum transform you have to set coefficients alpha.');
    end
    if(~h.y_is_set)
		error('Before doing a fastsum transform you have to set nodes y.');
    end
    
    fastsummex('trafo_direct',h.plan);
    h.f_is_set=true;
end %function

% Save and load %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function s = saveobj(h)
    s=struct('x',h.x,'y',h.y,'alpha',h.alpha,'d',h.d,'kernel',h.kernel,'c',h.c,...
        'flags',h.flags,'n',h.n,'p',h.p,'eps_I',h.eps_I,'eps_B',h.eps_B,...
        'nn_x',h.nn_x,'nn_y',h.nn_y,'m_x',h.m_x,'m_y',h.m_y);
end

end %methods

methods(Static)
function h = loadobj(s)
    h=fastsum(s.d,s.kernel,s.c,s.flags,s.n,s.p,s.eps_I,s.eps_B,...
        s.nn_x,s.m_x,s.nn_y,s.m_y);
    if ~isempty(s.x)
        h.x=s.x;
    end
    if ~isempty(s.alpha)
        h.alpha=s.alpha;
    end
    if ~isempty(s.y)
        h.y=s.y;
    end
end
end %methods
end %classdef

File: julia/fastsum/fastsum.jl
module fastsum

ending = ".so"

if Sys.iswindows( )
	ending = ".dll"
elseif Sys.isapple()
	ending = ".dylib"
end

const lib_path = string( @__DIR__, "/libfastsumjulia", ending )

mutable struct fastsum_plan
end

mutable struct Plan
	d::Integer					# dimension
	N::Integer					# number of source nodes
	M::Integer					# number of target nodes
	n::Integer					# expansion degree
	p::Integer					# degree of smoothness
	kernel::String				# name of kernel
	c::Vector{Float64}	# kernel parameters
	eps_I::Real					# inner boundary
	eps_B::Real				# outer boundary
	nn_x::Integer				# oversampled nn in x
	nn_y::Integer				# oversampled nn in y
	m_x::Integer				# NFFT-cutoff in x
	m_y::Integer				# NFFT-cutoff in y
	init_done::Bool			# bool for plan init
	finalized::Bool			# bool for finalizer
	flags::UInt32				# flags
	x::Ref{Float64}			# source nodes
	y::Ref{Float64}			# target nodes
	alpha::Ref{ComplexF64}		# source coefficients
	f::Ref{ComplexF64}				# target evaluations
	plan::Ref{fastsum_plan}		# plan (C pointer)
	
	function Plan( d::Integer, N::Integer, M::Integer, n::Integer, p::Integer, kernel::String, c::Vector{<:Real}, eps_I::Real, eps_B::Real, nn_x::Integer, nn_y::Integer, m_x::Integer, m_y::Integer, flags::UInt32 )
		new( d, N, M, n, p, kernel, Vector{Float64}(c), eps_I, eps_B, nn_x, nn_y, m_x, m_y, false, false, flags )
	end
end #struct fastsumplan

function Plan( d::Integer, N::Integer, M::Integer, n::Integer, p::Integer, kernel::String, c::Real, eps_I::Real, eps_B::Real, nn::Integer, m::Integer )

	if N <= 0
		error( "N has to be a positive Integer." )
	end
	
	if M <= 0
		error( "M has to be a positive Integer." )
	end
	
	if n <= 0
		error( "n has to be a positive Integer. ")
	end
	
	if m <= 0
		error( "m has to be a positive Integer." )
	end
	
	cv = Vector{Float64}( undef, 1 )
	cv[1] = Float64(c)
	
	Plan( d, N, M, n, p, kernel, cv, eps_I, eps_B, nn, nn, m, m, UInt32(0) )
	
end #constructor

function fastsum_init( p::Plan ) 

	ptr = ccall( ("jfastsum_alloc", lib_path), Ptr{fastsum_plan}, () )
	Core.setfield!( p, :plan, ptr )
	
	code = ccall( ("jfastsum_init",lib_path), Int64, (Ref{fastsum_plan}, Int32, Cstring, Ref{Float64}, UInt32, Int32, Int32, Float64, Float64, Int32, Int32, Int32, Int32, Int32, Int32), ptr, Int32(p.d), p.kernel, p.c, p.flags, Int32(p.n), Int32(p.p), Float64(p.eps_I), Float64(p.eps_B), Int32(p.N), Int32(p.M), Int32(p.nn_x), Int32(p.nn_y), Int32(p.m_x), Int32(p.m_y) )
	
	if code == 1
		error( "Unkown kernel." )
	end

	Core.setfield!( p, :init_done, true )
	finalizer( finalize_plan, p )
	
end #fastsum_init

function finalize_plan( p::Plan )

	if !p.init_done
		error( "Plan not initialized." )
	end

	if !p.finalized
		ccall( ("jfastsum_finalize",lib_path), Nothing, (Ref{fastsum_plan},), p.plan )
		Core.setfield!( p, :finalized, true )
	end

end #finalize_plan

function Base.setproperty!( p::Plan, v::Symbol, val )

	if !p.init_done
		fastsum_init( p )
	end

	if p.finalized
		error( "Plan already finalized" )
	end

	# edit source nodes
	if v == :x

		if p.d == 1
			if typeof(val) != Vector{Float64}
				error( "x has to be a Float64 vector." )
			end
			if size(val)[1] != p.N
				error( "x has to be a Float64 vector of length N." )
			end
	    else # => D >1
			if typeof(val) != Array{Float64, 2}
				error( "x has to be a Float64 matrix." )
			end
			if size(val)[1] != p.N || size(val)[2] != p.d
				error( "x has to be a Float64 matrix of size N." )
			end
		end

		ptr = ccall( ("jfastsum_set_x", lib_path), Ptr{Float64}, (Ref{fastsum_plan},Ref{Cdouble}), p.plan, val )
		Core.setfield!( p, v, ptr )

	 # edit target nodes
	 elseif v == :y

		if p.d == 1
			if typeof(val) != Vector{Float64}
				error( "y has to be a Float64 vector." )
			end
			if size(val)[1] != p.M
				error( "y has to be a Float64 vector of length M." )
			end
		else # => D > 1
			if typeof(val) != Array{Float64, 2}
				error( "y has to be a Float64 matrix." )
			end
			if size(val)[1] != p.M || size(val)[2] != p.d
				error( "y has to be a Float64 matrix of size M." )
			end
		end

		ptr = ccall( ("jfastsum_set_y", lib_path), Ptr{Float64}, (Ref{fastsum_plan},Ref{Cdouble}), p.plan, val )
		Core.setfield!( p, v, ptr )

   # edit source coefficients
	elseif v == :alpha

		if typeof(val) != Vector{ComplexF64}
			error( "alpha has to be a ComplexF64 vector." )
		end
		if size(val)[1] != p.N
			error( "alpha has to be a ComplexF64 vector of length N." )
		end

		ptr = ccall( ("jfastsum_set_alpha", lib_path), Ptr{ComplexF64}, (Ref{fastsum_plan},Ref{ComplexF64}), p.plan, val )
		Core.setfield!( p, v, ptr )

	elseif v == :M
	  	@warn("You can't modify the number of target nodes.")
  	elseif v == :N
	  	@warn("You can't modify the number of source nodes.")
  	elseif v == :n
	  	@warn("You can't modify the expansion degree.")
  	elseif v == :m
	  	@warn("You can't modify the cut-off parameter.")
  	elseif v == :p
	  	@warn("You can't modify the degree of smoothness.")
  	elseif v == :kernel
    	@warn("You can't modify the kernel.")
  	elseif v == :c
    	@warn("You can't modify the kernel parameters.")
  	elseif v == :eps_I
    	@warn("You can't modify the inner boundary.")
  	elseif v == :eps_B
    	@warn("You can't modify the outer boundary.")
  	elseif v == :plan
    	@warn("You can't modify the pointer to the fastsum plan.")
  	else
    	Core.setfield!( p, v, val )
  	end
  	
end # Base.setproperty!

# overwrite dot notation for plan struct in order to use C memory
function Base.getproperty( p::Plan, v::Symbol ) 

	if v == :x

		if !isdefined( p, :x )
			error("x is not set.")
		end

		ptr = Core.getfield( p, :x )

		if p.d == 1
			return unsafe_wrap( Vector{Float64}, ptr, p.N ) 				# get source nodes from C memory and convert to Julia type
		else
			return unsafe_wrap( Matrix{Float64}, ptr, (p.d, p.N) )  	# get source nodes from C memory and convert to Julia type
		end

	elseif v == :y

		if !isdefined( p, :y )
			error("y is not set.")
		end

		ptr = Core.getfield( p, :y )

		if p.d == 1
			return unsafe_wrap( Vector{Float64}, ptr, p.M )             
		else
			return unsafe_wrap( Matrix{Float64}, ptr, (p.d, p.M) )
		end 

	elseif v == :alpha

		if !isdefined( p, :alpha )
			error("alpha is not set.")
		end

		ptr = Core.getfield( p, :alpha )
		return unsafe_wrap( Vector{ComplexF64}, ptr, p.N ) 			# get coefficients from C memory and convert to Julia type

	elseif v == :f

		if !isdefined( p, :f )
	    error("f is not set.")
		end

		ptr = Core.getfield( p, :f )
		return unsafe_wrap( Vector{ComplexF64}, ptr, p.M )  			# get function values from C memory and convert to Julia type

	else
		return Core.getfield( p, v )
	end
end # Base.getproperty

function trafo( p::Plan )

	if p.finalized
		error( "Plan already finalized." )
	end

	if !isdefined( p, :x )
		error( "x has not been set." )
	end

	if !isdefined( p, :y )
		error( "y has not been set." )
	end

	if !isdefined( p, :alpha )
		error( "alpha has not been set." )
	end

	ptr = ccall( ( "jfastsum_trafo", lib_path), Ptr{ComplexF64}, (Ref{fastsum_plan},), p.plan )
	Core.setfield!( p, :f, ptr )

end #trafo

function trafo_exact( p::Plan )

	if p.finalized
		error( "Plan already finalized." )
	end

	if !isdefined( p, :x )
		error( "x has not been set." )
	end

	if !isdefined( p, :y )
		error( "y has not been set." )
	end

	if !isdefined( p, :alpha )
		error( "alpha has not been set." )
	end

	ptr = ccall( ("jfastsum_exact", lib_path), Ptr{ComplexF64}, (Ref{fastsum_plan},), p.plan )
	Core.setfield!( p, :f, ptr )

end #trafo

function get_num_threads()
	return ccall(("nfft_get_num_threads",lib_path),Int64,())
end

function set_num_threads(nthreads::Number)
	ccall(("nfft_set_num_threads",lib_path),Nothing,(Int64,),convert(Int64,nthreads))
end

end #module
